File: p4179.cpp
#include<bits/stdc++.h>
using namespace std;
#define MAX 1002

vector<vector<char>> MAP(MAX, vector<char>(MAX, 'E'));
vector<vector<int>> visited(MAX, vector<int>(MAX, 0));
int N, M, cnt;
int dx[] = {1, 0, -1, 0};
int dy[] = {0, 1, 0, -1};
queue<pair<int, int>> firePos, jhPos;
int ex = -1, ey = -1;

void goFire(){
    queue<pair<int, int>> tmp;
    while(!firePos.empty()){
        int x = firePos.front().second;
        int y = firePos.front().first;
        firePos.pop();

        for(int i = 0; i < 4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(nx < 1 || nx >= M + 1 || ny < 1 || ny >= N + 1) continue;
            if(MAP[ny][nx] == '.') {
                MAP[ny][nx] = 'F';
                tmp.push(make_pair(ny, nx));
            }
        }
    }
    firePos = tmp;
}

void goJH(){
    queue<pair<int, int>> tmp;
    while(!jhPos.empty()){
        int x = jhPos.front().second;
        int y = jhPos.front().first;
        jhPos.pop();

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || nx >= M + 2 || ny < 0 || ny >= N + 2) continue;
            if(MAP[ny][nx] == 'E'){
                ex = x;
                ey = y;
                return;
            }
            if (visited[ny][nx] == 0 && MAP[ny][nx] == '.'){
                visited[ny][nx] = visited[y][x] + 1;
                tmp.push(make_pair(ny, nx));
            }
        }
    }
    jhPos = tmp;
}

int main(){
    cin >> N >> M;
    for(int i = 1; i <= N; i++){
        for(int j = 1; j <= M; j++){
            cin >> MAP[i][j];
            if(MAP[i][j] == 'J'){
                jhPos.push(make_pair(i, j));
                visited[i][j] = 0;
            }
            else if(MAP[i][j] == 'F'){
                firePos.push(make_pair(i, j));
            }
        }
    }
    while(!jhPos.empty()){
        goFire();
        goJH();
        if(ex != -1 && ey != -1) break;
    }

    if(ex != -1 && ey != -1){
        cout << visited[ey][ex] + 1 << "\n";
    }
    else cout << "IMPOSSIBLE\n";
}

File: p4179.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class p4179 {
    public static int N, M, flag;
    public static int[] dy = {0, 1, 0, -1};
    public static int[] dx = {1, 0, -1, 0};
    public static boolean[][] visited;
    public static char[][] map;
    public static Queue<int[]> firePos = new LinkedList<>();;
    public static Queue<int[]> jhPos = new LinkedList<>();;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        N = Integer.parseInt(input[0]);
        M = Integer.parseInt(input[1]);
        map = new char[N][M];
        visited = new boolean[N][M];

        for(int i = 0; i < N; i++){
            map[i] = br.readLine().toCharArray();
            
            for(int j = 0; j < M; j++){
                if(map[i][j] == 'J') {
                    jhPos.add(new int[]{i, j});
                    visited[i][j] = true;
                }
                else if(map[i][j] == 'F') {
                    firePos.add(new int[]{i, j});
                }
            }
        }

        int res = 0;
        while(flag == 0){
            goJH();
            goFire();
            res++;
        }

        if(flag == 1) System.out.println(res);
        else if(flag == 2) System.out.println("IMPOSSIBLE");
    }

    public static void goJH(){
        Queue<int[]> tmp = new LinkedList<>();

        while(!jhPos.isEmpty()){
            int y = jhPos.peek()[0];
            int x = jhPos.peek()[1];
            jhPos.poll();

            if(map[y][x] == 'F') continue;

            for(int i = 0; i < 4; i++){
                int ny = y + dy[i];
                int nx = x + dx[i];

                if(ny < 0 || nx < 0 || ny >= N || nx >= M){
                    flag = 1;
                    return;
                }
                if(visited[ny][nx] || map[ny][nx] != '.') continue;
                visited[ny][nx] = true;
                tmp.add(new int[]{ny , nx});
            }
        }

        jhPos = tmp;
        if(jhPos.isEmpty()) flag = 2;
    }

    public static void goFire(){
        Queue<int[]> tmp = new LinkedList<>();

        while(!firePos.isEmpty()){
            int y = firePos.peek()[0];
            int x = firePos.peek()[1];
            firePos.poll();

            for(int i = 0; i < 4; i++){
                int ny = y + dy[i];
                int nx = x + dx[i];

                if(ny < 0 || nx < 0 || ny >= N || nx >= M) continue;
                if(map[ny][nx] == '.'){
                    map[ny][nx] = 'F';
                    tmp.add(new int[]{ny, nx});
                }
            }
        }

        firePos = tmp;
    }
}